### PR TITLE
feat: register enterprise cluster

### PR DESCRIPTION
**Pull request by pawelpaszki**
## Description
feat: register enterprise cluster

## Verification Steps
run the code and hit http://localhost:8000/api/kafkas_mgmt/v1/clusters (POST) endpoint. make sure that valid cluster id (alphanumeric with 32 chars length), external cluster id (in the correct format) and dns name (also in the correct format) are provided in the request. Cluster should be registered and in `accepted` state.

Then cluster manager should progress the cluster through various states up until `waiting_for_kas_fleetshard_operator`

tests will follow after initial code review is done

## Checklist (Definition of Done)

- [ ] All acceptance criteria specified in JIRA have been completed
- [ ] Unit and integration tests added that prove the fix is effective or the feature works (tested against emulated and non-emulated OCM environment)
- [ ] Documentation added for the feature
- [ ] CI and all relevant tests are passing
- [ ] Code Review completed
- [ ] Verified independently by reviewer
- [ ] All PR comments are resolved either by addressing them or creating follow up tasks
- [ ] Required metrics/dashboards/alerts have been added (or PR created).
- [ ] Required Standard Operating Procedure (SOP) is added.
- [ ] JIRA has been created for changes required on the client side

### CODE REVIEW

File: internal/kafka/internal/workers/cluster_mgrs/deprovisioning_clusters_mgr.go
Comment (machi1990) [inline internal/kafka/internal/workers/cluster_mgrs/deprovisioning_clusters_mgr.go:110]: We've to add a logic in this reconciler to skip deprovisioning of enterprise clusters.

Additionally, we'll need a similar kind of logic for dynamic scale up/down reconcilers as well as we absolute shouldn't exercise any dynamic scaling logic for enterprise clusters. That could follow up in a separate JIRA.

File: openapi/kas-fleet-manager.yaml
Comment (machi1990) [inline openapi/kas-fleet-manager.yaml:1653]: What do you think of separating out the `parameters` into two arrays?
i.e `fleetshard_parameters` and `observability_stack_parameters`?

File: internal/kafka/internal/handlers/cluster.go
Comment (machi1990) [inline internal/kafka/internal/handlers/cluster.go:71]: I am wondering how does this logic interplay with `AccessControlListMiddleware` logic: this middleware controls access to all the resources/endpoints.

I think it won't fair well. 
e.g 
1. if access control is disabled, then no access to any endpoint is granted meaning everyone except for those listed in the list, are locked out

And thus I propose we've a complete and yet "temporary" separate configuration (until we've quota in place) to control only access to the enterprise clusters registration endpoint.

File: internal/kafka/internal/presenters/cluster.go
Comment (machi1990) [inline internal/kafka/internal/presenters/cluster.go:55]: Let me know what you think of https://github.com/bf2fc6cc711aee1a0c2a/kas-fleet-manager/pull/1429#discussion_r1039495248

File: internal/kafka/internal/workers/cluster_mgrs/deprovisioning_clusters_mgr.go
Comment (pawelpaszki) [reply internal/kafka/internal/workers/cluster_mgrs/deprovisioning_clusters_mgr.go:110]: for dynamic scaling - the calculations exclude enterprise clusters (see https://github.com/bf2fc6cc711aee1a0c2a/kas-fleet-manager/pull/1429/files#diff-c1b9a697a5a30e72d8cc46881d4360947c26e2b190bbb2cf65c8f52651cfe773R741)

File: internal/kafka/internal/presenters/cluster.go
Comment (pawelpaszki) [reply internal/kafka/internal/presenters/cluster.go:55]: sure. I can separate them. btw. one thing I noticed is inconsistency in the id names format, e.g.:

```

{
    "cluster_id": "****",
    "status": "cluster_accepted",
    "parameters": [
        {
            "id": "gateway",
            "value": "****"
        },
        {
            "id": "tenant",
            "value": "****"
        },
        {
            "id": "redHatSsoAuthServerUrl",
            "value": "****"
        },
        {
            "id": "redHatSsoRealm",
            "value": "****"
        },
        {
            "id": "metricsClientId",
            "value": "****"
        },
        {
            "id": "metricsSecret",
            "value": "****"
        },
        {
            "id": "sso-auth-server-url",
            "value": "****"
        },
        {
            "id": "sso-client-id",
            "value": "****"
        },
        {
            "id": "sso-secret",
            "value": "****"
        },
        {
            "id": "control-plane-url",
            "value": "****"
        },
        {
            "id": "cluster-id",
            "value": "****"
        },
        {
            "id": "poll-interval",
            "value": "****"
        },
        {
            "id": "resync-interval",
            "value": "****"
        }
    ]
}

```

File: internal/kafka/internal/workers/cluster_mgrs/deprovisioning_clusters_mgr.go
Comment (machi1990) [reply internal/kafka/internal/workers/cluster_mgrs/deprovisioning_clusters_mgr.go:110]: Ah apparently, I had missed that. Nice work, thanks!

File: internal/kafka/internal/presenters/cluster.go
Comment (machi1990) [reply internal/kafka/internal/presenters/cluster.go:55]: > one thing I noticed is inconsistency in the id names format,

let's keep the id names as they are since the parameters are for two separate Operators: Fleetshard and Observability. 
So they'll be two separate secrets where the id will be the `key` in the secret and each Operator expect them in different formats.

File: internal/kafka/internal/workers/cluster_mgrs/clusters_mgr.go
Comment (machi1990) [inline internal/kafka/internal/workers/cluster_mgrs/clusters_mgr.go:230]: Thanks for adding this skip logic:

For now what we are missing is a database update to move the status of the cluster to the next phase 
e.g when the enterprise cluster is in accepted state, the only thing we'll do is move it to the cluster_provisioning state.

And in the method below
```go
processProvisioningClusters,
```

when the enterprise cluster is in provisioning state, the only thing we'll do is move it to `cluster_provisioned` state.

Lastly, in this method  
```go
processProvisionedClusters
```

the enterprise cluster, won't be terraformed but moved to the waiting_for_kas_fleetshard_operator state.

File: internal/kafka/internal/workers/cluster_mgrs/clusters_mgr.go
Comment (machi1990) [inline internal/kafka/internal/workers/cluster_mgrs/clusters_mgr.go:258]: Let's take the comment in https://github.com/bf2fc6cc711aee1a0c2a/kas-fleet-manager/pull/1429/files#r1039519780 onto account here

File: internal/kafka/internal/workers/cluster_mgrs/clusters_mgr.go
Comment (machi1990) [inline internal/kafka/internal/workers/cluster_mgrs/clusters_mgr.go:288]: let's take the comment in https://github.com/bf2fc6cc711aee1a0c2a/kas-fleet-manager/pull/1429/files#r1039519780 onto account

**Review by machi1990 — COMMENTED** [review]
Thanks @pawelpaszki 

Great work, I've left some comments. Let me know what you think.

File: internal/kafka/internal/workers/cluster_mgrs/deprovisioning_clusters_mgr.go
Comment (pawelpaszki) [reply internal/kafka/internal/workers/cluster_mgrs/deprovisioning_clusters_mgr.go:110]: for the first bullet point - clusters returned in the function follows into `err = c.ClusterService.UpdateMultiClusterStatus(idsOfClustersToDeprovision, api.ClusterDeprovisioning)` are excluding cluster with `cluster_type` being `enterprise`

File: internal/kafka/internal/handlers/cluster.go
Comment (machi1990) [reply internal/kafka/internal/handlers/cluster.go]: thank you!

**Comment by pawelpaszki** [comment]
@miguelsorianod - can you take a look at the changes wrt dynamic scaling please?

File: pkg/api/cluster_types.go
Comment (machi1990) [inline pkg/api/cluster_types.go:178]:
```suggestion
	OrganisationID      string `json:"organisation_id"`
```

File: internal/kafka/internal/handlers/cluster.go
Comment (machi1990) [inline internal/kafka/internal/handlers/cluster.go:75]:
```suggestion
				OrganisationID:      orgId,
```

File: internal/kafka/internal/handlers/cluster.go
Comment (pb82) [inline internal/kafka/internal/handlers/cluster.go:73]: Is accepted the right status? If we're not reconciling the cluster, it won't ever reach another status.

Comment (pb82) [reply internal/kafka/internal/handlers/cluster.go:73]: nvw. I see that you are updating the status in the reconcilers

File: internal/kafka/internal/presenters/cluster.go
Comment (pb82) [inline internal/kafka/internal/presenters/cluster.go:12]: @machi1990 @pawelpaszki if we're going to install the Observability Operator through SyncSets even for Enterprise, we probably don't need to return this.

Comment (machi1990) [reply internal/kafka/internal/presenters/cluster.go:12]: I agree.

Comment (pawelpaszki) [reply internal/kafka/internal/presenters/cluster.go]: removed

**Review by machi1990 — COMMENTED** [review]
the code change lgtm

Thanks @pawelpaszki @pb82

**Comment by pawelpaszki** [comment]
> the code change lgtm
> 
> Thanks @pawelpaszki @pb82

thanks Manyanda and Peter. I am working on tests now

File: internal/kafka/internal/workers/cluster_mgrs/clusters_mgr.go
Comment (ziccardi) [inline internal/kafka/internal/workers/cluster_mgrs/clusters_mgr.go:234]: `continue` is useless here

Comment (ziccardi) [inline internal/kafka/internal/workers/cluster_mgrs/clusters_mgr.go:234]: is the `continue` here just to skip the log in case of errors? If yes, I believe an `else` would be cleaner

File: internal/kafka/internal/workers/cluster_mgrs/clusters_mgr.go
Comment (ziccardi) [inline internal/kafka/internal/workers/cluster_mgrs/clusters_mgr.go:261]: Same as for the previous comment: this `continue` is not skipping anything and can be removed

File: internal/kafka/internal/workers/cluster_mgrs/clusters_mgr.go
Comment (ziccardi) [inline internal/kafka/internal/workers/cluster_mgrs/clusters_mgr.go:268]: Same comment as before: if this `continue` is here just to skip the log, it can be replaced with an `else`

File: internal/kafka/internal/workers/cluster_mgrs/clusters_mgr.go
Comment (ziccardi) [inline internal/kafka/internal/workers/cluster_mgrs/clusters_mgr.go:297]: see previous comments

File: internal/kafka/internal/workers/cluster_mgrs/clusters_mgr.go
Comment (ziccardi) [inline internal/kafka/internal/workers/cluster_mgrs/clusters_mgr.go:304]: see previous comments

File: internal/kafka/internal/workers/cluster_mgrs/clusters_mgr.go
Comment (ziccardi) [inline internal/kafka/internal/workers/cluster_mgrs/clusters_mgr.go:331]: see previous comments

File: internal/kafka/internal/workers/cluster_mgrs/clusters_mgr_test.go
Comment (ziccardi) [inline internal/kafka/internal/workers/cluster_mgrs/clusters_mgr_test.go:822]: Should be deleted?

File: openapi/kas-fleet-manager.yaml
Comment (machi1990) [inline openapi/kas-fleet-manager.yaml:1630]:
```suggestion
        - cluster_dns_name
```

File: openapi/kas-fleet-manager.yaml
Comment (machi1990) [inline openapi/kas-fleet-manager.yaml:1641]:
```suggestion
        cluster_ingress_dns_name:
          description: dns name of the cluster. Can be obtained from the response JSON of the /api/clusters_mgmt/v1/clusters/<cluster_id>/ingresses (dns_name)
          type: string
```

I am wondering if we should explicit name this field the way I am suggesting? As it is about ingresses. What do you think?

File: internal/kafka/internal/workers/cluster_mgrs/clusters_mgr_test.go
Comment (pawelpaszki) [reply internal/kafka/internal/workers/cluster_mgrs/clusters_mgr_test.go:822]: I commented it out when testing. the comments should and will be removed

File: internal/kafka/internal/handlers/cluster.go
Comment (machi1990) [inline internal/kafka/internal/handlers/cluster.go:66]: What do you think about extracting this logic quota onto a middleware that can be used only for this handler? 
Later on the middleware will deal with quota checks and the checks are performed before even landing to the handler

Comment (pawelpaszki) [reply internal/kafka/internal/handlers/cluster.go:66]: do you mean something like `AccessControlListMiddleware`?

Comment (machi1990) [reply internal/kafka/internal/handlers/cluster.go:66]: Yes. Along those lines. But the new middleware will only be applied for this new endpoint.

File: internal/kafka/internal/services/clusters.go
Comment (machi1990) [inline internal/kafka/internal/services/clusters.go:741]: The implication of this exclusion is that:
1. enterprise data plane clusters won't be considering in dynamic scaling evaluation, this is fine.
2. But it also means that we won't have streaming unit metrics for enterprise data planes. I feel like this is something that might be needed. However, even the capacity used/max metrics for enterprise data plane might need to be revisited slightly a little. So I am happy with us capturing a JIRA about this so that we do not forget.

@pawelpaszki @pb82

Comment (pawelpaszki) [reply internal/kafka/internal/services/clusters.go:741]: follow up Jira makes sense to me in this case

File: internal/kafka/internal/routes/route_loader.go
Comment (machi1990) [inline internal/kafka/internal/routes/route_loader.go:59]: I think this can be removed

File: internal/kafka/internal/handlers/cluster_test.go
Comment (ziccardi) [inline internal/kafka/internal/handlers/cluster_test.go:290]: The `resp.Body` won't be closed when the test fails.

File: pkg/providers/core.go
Comment (machi1990) [inline pkg/providers/core.go:84]: The acl checking mechanism seems specific to kafka, I think we could move the middleware and the config out of the pkg/providers/core.go to https://github.com/bf2fc6cc711aee1a0c2a/kas-fleet-manager/blob/40fbcaa88a6bda9e82093f9f7774ab6180b562c2/internal/kafka/providers.go

File: internal/kafka/internal/services/clusters.go
Comment (machi1990) [reply internal/kafka/internal/services/clusters.go:741]: I'll create one. Thanks

File: pkg/providers/core.go
Comment (pawelpaszki) [reply pkg/providers/core.go]: thanks. moved it to internal folder

**Review by machi1990 — APPROVED** [review]
Verified this against a cluster. Works as expected. Great work, thanks @pawelpaszki

**Comment by pawelpaszki** [comment]
thanks Manyanda. I will merge this PR once CI passes and add tests for the acl stuff in the follow up PR

**Pull request merged**
